### PR TITLE
Hookup node settings screen.

### PR DESCRIPTION
**Pull request by abouolia**

### CODE REVIEW

File: packages/renderer/src/hooks/haveno/useSetMoneroNodeSettings.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// =============================================================================
+
+import { QueryKeys } from "@constants/query-keys";
+import { useMutation, useQueryClient } from "react-query";
+import { useHavenoClient } from "./useHavenoClient";
+
+interface SetMeneroNodeSettingsQueryData {
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useSetMoneroNodeSettings.ts]: Code convention: we use the pattern `SomethingVariables` or simply, `Variables`. [Here's an example](https://github.com/Viabl-Ventures/haveno-ui/blob/develop/packages/renderer/src/hooks/haveno/useCreateAccount.ts#L20)

File: packages/renderer/src/hooks/haveno/useStartMoneroNode.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+export function useStartMoneroNode() {
+  const client = useHavenoClient();
+  const queryClient = useQueryClient();
+
+  return useMutation<void, Error, MoneroNodeSettings>(
+    (data: MoneroNodeSettings) => client.startMoneroNode(data),
+    {
+      onSuccess: () => {
+        queryClient.invalidateQueries(QueryKeys.MoneroNodeIsRunning);
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useStartMoneroNode.ts:30]: Nice!

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx [hunk 3]
```diff
@@ -14,28 +14,70 @@
...
+import { FormattedMessage, useIntl } from "react-intl";
 import { LangKeys } from "@constants/lang";
 import { TextInput } from "@atoms/TextInput";
+import { useMoneroNodeSettings } from "@hooks/haveno/useMoneroNodeSettings";
+import { useSetMoneroNodeSettings } from "@hooks/haveno/useSetMoneroNodeSettings";
+import { NodeLocalStopDeamon } from "./NodeLocalStopDeamon";
+import type { NodeLocalFormValues } from "./_hooks";
+import { useNodeLocalFormValidation } from "./_hooks";
+import { transformSettingsRequestToForm } from "./_utils";
+import { showNotification } from "@mantine/notifications";
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx]: Minor nitpick: vendor imports go first

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx [hunk 4]
```diff
@@ -14,28 +14,70 @@
...
+
+  const validation = useNodeLocalFormValidation();
+
+  const form = useForm<NodeLocalFormValues>({
     initialValues: {
       blockchainLocation: "",
       startupFlags: "",
       deamonAddress: "",
       port: "",
+      ...(nodeSettings
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx:44]: Nice!

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalStopDeamon.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        showNotification({
+          color: "red",
+          message: err.message,
+          title: "Something went wrong",
+        });
+      });
+  };
+  // Handle the start button click.
+  const handleStartBtnClick = () => {
+    if (nodeSettings) {
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalStopDeamon.tsx]: Use early return for better code readability:
```ts
if(!nodeSettings) {
  return;
}
```

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalStopDeamon.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            color: "red",
+            message: err.message,
+            title: "Something went wrong",
+          });
+        });
+    }
+  };
+
+  return (
+    <div className={classes.actions}>
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalStopDeamon.tsx:84]: `Box` preferred over `div`

File: packages/renderer/src/pages/Account/NodeSettings/_utils.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    startupFlags: nodeSettings?.startupFlagsList.join(", ") || "",
+    deamonAddress: transfromBootstrapUrl(nodeSettings?.bootstrapUrl || ""),
+    port: transformPort(nodeSettings?.bootstrapUrl || ""),
+  };
+}
+
+function transformPort(urlAsString: string) {
+  try {
+    const url = new URL(urlAsString);
+    return url.port;
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/_utils.ts:39]: Neat! Does this handle malformed URLs?

File: packages/renderer/src/hooks/haveno/useIsMoneroNodeRunning.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// =============================================================================
+
+import { QueryKeys } from "@constants/query-keys";
+import { useQuery } from "react-query";
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useIsMoneroNodeRunning.ts:17]: Minor nitpick: vendor imports first

File: packages/renderer/src/hooks/haveno/useMoneroNodeSettings.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// =============================================================================
+
+import { QueryKeys } from "@constants/query-keys";
+import type { MoneroNodeSettings } from "haveno-ts";
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useMoneroNodeSettings.ts:18]: Minor nitpick: vendor imports first

File: packages/renderer/src/hooks/haveno/useIsMoneroNodeRunning.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import { useHavenoClient } from "./useHavenoClient";
+
+export function useIsMoneroNodeRunning() {
+  const client = useHavenoClient();
+
+  return useQuery<boolean, Error>(QueryKeys.MoneroNodeIsRunning, async () => {
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useIsMoneroNodeRunning.ts]: The arrow function can be condensed to a one-liner:
```ts
async () => client.isMoneroNodeRunning()
```

File: packages/renderer/src/hooks/haveno/useMoneroNodeSettings.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+import { useQuery } from "react-query";
+import { useHavenoClient } from "./useHavenoClient";
+
+export function useMoneroNodeSettings() {
+  const client = useHavenoClient();
+
+  return useQuery<MoneroNodeSettings | undefined, Error>(
+    QueryKeys.MoneroNodeSettings,
+    async () => {
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useMoneroNodeSettings.ts:27]: Can be condensed to a one-liner

File: packages/renderer/src/hooks/haveno/useMoneroRemoteNodes.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// =============================================================================
+
+import { QueryKeys } from "@constants/query-keys";
+import { useQuery } from "react-query";
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useMoneroRemoteNodes.ts:17]: Minor nitpick: Vendor imports first

File: packages/renderer/src/hooks/haveno/useMoneroRemoteNodes.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  title: string;
+  isActive: boolean;
+}
+
+export function useMoneroRemoteNodes() {
+  // const client = useHavenoClient();
+
+  return useQuery<MoneroRemoteNodes[], Error>(
+    QueryKeys.MoneroRemoteNodes,
+    async () => {
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useMoneroRemoteNodes.ts:31]: Nice stub :)

File: packages/renderer/src/hooks/haveno/useSetMoneroNodeSettings.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// =============================================================================
+
+import { QueryKeys } from "@constants/query-keys";
+import { useMutation, useQueryClient } from "react-query";
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useSetMoneroNodeSettings.ts:17]: Minor nitpick: Vendor imports first

File: packages/renderer/src/hooks/haveno/useStartMoneroNode.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// =============================================================================
+
+import { QueryKeys } from "@constants/query-keys";
+import type { MoneroNodeSettings } from "haveno-ts";
```
Comment (schowdhuri) [inline packages/renderer/src/hooks/haveno/useStartMoneroNode.ts:18]: Minor nitpick: Vendor imports first

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx [hunk 16]
```diff
@@ -14,28 +14,70 @@
 //  limitations under the License.
 // =============================================================================
 
-import { Box, Stack, Grid, createStyles } from "@mantine/core";
-import { useForm } from "@mantine/form";
+import { Box, Stack, Grid, Group } from "@mantine/core";
+import { joiResolver, useForm } from "@mantine/form";
 import { Button } from "@atoms/Buttons";
-import { FormattedMessage } from "react-intl";
+import { FormattedMessage, useIntl } from "react-intl";
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx:19]: Minor nitpick: Vendor imports first

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx (hunk 3 above)
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx]: Minor nitpick: Vendor imports first

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx [hunk 17]
```diff
@@ -14,28 +14,70 @@
...
       port: "",
+      ...(nodeSettings
+        ? transformSettingsRequestToForm(nodeSettings.toObject())
+        : {}),
     },
+    validate: joiResolver(validation),
   });
 
+  const handleFormSubmit = (values: NodeLocalFormValues) => {
+    updateNodeSettings({
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx:52]: Good stuff!

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx [hunk 18]
```diff
@@ -67,6 +109,7 @@ export function NodeLocalForm() {
                     defaultMessage="Deamon Address"
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx:109]: Please change "Deamon" to "Daemon" everywhere in the codebase. I understand the confusion - it is spelled incorrectly in the Figma too :)

File: packages/renderer/src/pages/Account/NodeSettings/NodeRemoteStatus.tsx [hunk 19]
```diff
@@ -19,19 +19,22 @@ import { Button } from "@atoms/Buttons";
 import { NodeStatus, NodeStatusType } from "@atoms/NodeStatus";
 import { FormattedMessage } from "react-intl";
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/NodeRemoteStatus.tsx]: Minor nitpick: Vendor imports first

File: packages/renderer/src/pages/Account/NodeSettings/_hooks.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+import * as Joi from "joi";
+
+export interface NodeLocalFormValues {
+  blockchainLocation: string;
+  startupFlags: string;
+  deamonAddress: string;
+  port: string;
+}
+
+export const useNodeLocalFormValidation = () => {
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/_hooks.ts]: Please change to a `function`

File: packages/renderer/src/pages/Account/NodeSettings/_utils.ts [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+
+import type { MoneroNodeSettings } from "haveno-ts";
+import type { NodeLocalFormValues } from "./_hooks";
+
+/**
+ * Transformes the settings request values to form.
+ * @param    {MoneroNodeSettings.AsObject} nodeSettings
+ * @returns  {NodeLocalFormValues}
+ */
+export function transformSettingsRequestToForm(
```
Comment (schowdhuri) [inline packages/renderer/src/pages/Account/NodeSettings/_utils.ts:25]: 🔥

**Review by schowdhuri — CHANGES_REQUESTED** [review]
Super cool, thank you! Added my comments - minor style nitpicks which should take only a few minutes to address :)

PS: Going forward all PRs must contain tests and storybook entries

File: packages/renderer/src/hooks/haveno/useStartMoneroNode.ts (hunk 2 above)
Comment (abouolia) [reply packages/renderer/src/hooks/haveno/useStartMoneroNode.ts:30]: Dude, My hands are dirty with React query.

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalStopDeamon.tsx (hunk 5 above)
Comment (abouolia) [reply packages/renderer/src/pages/Account/NodeSettings/NodeLocalStopDeamon.tsx]: You got me :D

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx (hunk 18 above)
Comment (abouolia) [reply packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx:109]: At least I learned new vocabulary. 😄

File: packages/renderer/src/hooks/haveno/useSetMoneroNodeSettings.ts (hunk 1 above)
Comment (abouolia) [reply packages/renderer/src/hooks/haveno/useSetMoneroNodeSettings.ts]: Fixed.

File: packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx (hunk 3 above)
Comment (abouolia) [reply packages/renderer/src/pages/Account/NodeSettings/NodeLocalForm.tsx]: All fixed.

**Review by schowdhuri — APPROVED** [review]
🚀

**Pull request merged**